### PR TITLE
implemented unleaning feature in inverted_index

### LINKED ISSUE #51: Support unlearning for `inverted_index`
`inverted_index` is stable and easy to use. But, it doesn't support unlearning. Users who need unlearning cannot choose `inverted_index`.

I have two ideas to achieve it:

- implement `inverted_index` nearest neighbor module, and implement unlearning
- implement unlearning in `inverted_index` recommender

The former is preferable.

### PR DESCRIPTION
**Pull request by kumagi**
`inverted_index` is tend to consume memory, there is big benefit for supporting `unlearner` in inverted index. fix #51

### CODE REVIEW

File: jubatus/core/driver/recommender_test.cpp [hunk 1]
```diff
@@ -156,6 +157,111 @@ TEST_P(nn_recommender_test, update) {
...
+}
+
+TEST(inverted_index_unlearner, lru_delete) {
+  unlearner::lru_unlearner::config conf;
+  conf.max_size = 3;
+  shared_ptr<unlearner::unlearner_base> unl(new unlearner::lru_unlearner(conf));
+  shared_ptr<driver::recommender> recommender =
+      shared_ptr<driver::recommender>(new driver::recommender(
+        shared_ptr<recommender_base>(
+            new core::recommender::inverted_index(unl)),
```
Comment (gwtnb) [inline jubatus/core/driver/recommender_test.cpp]: This may cause memory leak.

File: jubatus/core/driver/recommender_test.cpp [hunk 2]
```diff
@@ -156,6 +157,111 @@ TEST_P(nn_recommender_test, update) {
...
+}
+
+TEST(inverted_index_unlearner, mix) {
+  unlearner::lru_unlearner::config conf;
+  conf.max_size = 3;
+  shared_ptr<unlearner::unlearner_base> unl1(new unlearner::lru_unlearner(conf));
+  shared_ptr<driver::recommender> recommender1 =
+      shared_ptr<driver::recommender>(new driver::recommender(
+        shared_ptr<recommender_base>(
+            new core::recommender::inverted_index(unl1)),
```
Comment (gwtnb) [inline jubatus/core/driver/recommender_test.cpp]: same as above.

File: jubatus/core/driver/recommender_test.cpp [hunk 3]
```diff
@@ -156,6 +157,111 @@ TEST_P(nn_recommender_test, update) {
...
+        make_tf_idf_fv_converter()));
+  framework::linear_mixable* mixable1 =
+      dynamic_cast<framework::linear_mixable*>(recommender1->get_mixable());
```
Comment (gwtnb) [inline jubatus/core/driver/recommender_test.cpp]: check null.

File: jubatus/core/driver/recommender_test.cpp [hunk 4]
```diff
@@ -156,6 +157,111 @@ TEST_P(nn_recommender_test, update) {
...
+  recommender1->update_row("id1", create_datum_str("a", "a b c"));
+  recommender1->update_row("id2", create_datum_str("a", "d e f"));
+  recommender1->update_row("id3", create_datum_str("a", "e f g"));
+
+  shared_ptr<unlearner::unlearner_base> unl2(new unlearner::lru_unlearner(conf));
+  shared_ptr<driver::recommender> recommender2 =
+      shared_ptr<driver::recommender>(new driver::recommender(
+        shared_ptr<recommender_base>(
+            new core::recommender::inverted_index(unl2)),
```
Comment (gwtnb) [inline jubatus/core/driver/recommender_test.cpp]: same as above.

File: jubatus/core/driver/recommender_test.cpp [hunk 5]
```diff
@@ -156,6 +157,111 @@ TEST_P(nn_recommender_test, update) {
...
+        make_tf_idf_fv_converter()));
+  recommender2->update_row("id2", create_datum_str("a", "d e f"));
+  recommender2->update_row("id3", create_datum_str("a", "e f g"));
+  recommender2->update_row("id4", create_datum_str("a", "f g h"));
+  framework::linear_mixable* mixable2 =
+      dynamic_cast<framework::linear_mixable*>(recommender2->get_mixable());
```
Comment (gwtnb) [inline jubatus/core/driver/recommender_test.cpp]: check null.

**Comment by kumagi** [comment]
https://gist.github.com/kumagi/a9a1815ccf7d7059197f
example of configure file

File: jubatus/core/driver/recommender_test.cpp (hunk 1 above)
Comment (kumagi) [inline jubatus/core/driver/recommender_test.cpp]: OK.

**Comment by kmaehashi** [comment]
unlearner's touch operation may fail when `sticky_pattern` is specified and number of IDs that matches `sticky_pattern` equals to `max_size`.

classifier avoids this by calling `can_touch` before updating the model.

https://github.com/jubatus/jubatus_core/blob/0.1.0/jubatus/core/classifier/perceptron.cpp#L31
https://github.com/jubatus/jubatus_core/blob/0.1.0/jubatus/core/classifier/linear_classifier.cpp#L223

**Comment by kumagi** [comment]
@kmaehashi OK, It will updates only touchable values.

**Pull request merged**